GitCommands/Git/RevisionDiffProvider.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace GitCommands
+{
+    /// <summary>
+    /// Translate GitRevision including artificial commits to diff options
+    /// Closely related to GitRevision.cs 
+    /// </summary>
+    public static class RevisionDiffProvider
+    {
+        private const string StagedOpt = "--cached";
+
+        /// <summary>
+        /// Translate the revision string to an option string
+        /// Artificial "commits" are options, handle aliases too
+        /// (order and handling of empty arguments is not handled here)
+        /// </summary>
+        /// <param name="rev"></param>
+        /// <returns></returns>
+        private static string ArtificialToDiffOptions(string rev)
+        {
+            if (rev.IsNullOrEmpty() || rev == GitRevision.UnstagedGuid) { rev = string.Empty; }
+            else if (rev == "^" || rev == GitRevision.UnstagedGuid + "^" || rev == GitRevision.IndexGuid) { rev = StagedOpt; }
+            else
+            {
+                //Normal commit
+                if (rev == "^^" || rev == GitRevision.UnstagedGuid + "^^" || rev == GitRevision.IndexGuid + "^") { rev = "HEAD"; }
+                else if (rev == "^^^" || rev == GitRevision.UnstagedGuid + "^^^" || rev == GitRevision.IndexGuid + "^^") { rev = "HEAD^"; }
+                rev = rev.QuoteNE();
+            }
+
+            return rev;
+        }
+
+        /// <summary>
+        /// options to git-diff from GE arguments, including artificial commits
+        /// </summary>
+        /// <param name="from">The first revision</param>
+        /// <param name="to">The second "current" revision</param>
+        /// <returns></returns>
+        public static string Get(string from, string to)
+        {
+            string extra = string.Empty;
+            from = ArtificialToDiffOptions(from);
+            to = ArtificialToDiffOptions(to);
+
+            //Note: As artificial are options, diff unstage..unstage and 
+            // stage..stage will show output, different from e.g. HEAD..HEAD
+            //Diff-to-itself is not always disabled why this is not handled as error in release builds
+            Debug.Assert(!(from == to && (from.IsNullOrEmpty() || from == StagedOpt)),
+                "Unexpectedly two identical artificial revisions to diff: " + from +
+                ". This will be displayed as diff to HEAD, not an identical diff.");
```

Comment:
Should it throw instead?
If it happens in release build, what would be the expected behavior?

Reply:
No throw. Comparing unstaged to unstaged or staged to staged cannot be represented in git-diff arguments. It would have been easier also from other perspectives if Git always handled staged/unstaged as commits, but it will not happen...

The UI tries to block these comparisons, but it is not completely possible in all situations (I assume). An exception would be impropriate as this is not a programming error and not really a user error either. A popup is not possible in this class and reporting it back to the caller is an option but will add maybe 10 lines of code in the 4 places it is called. I did not bother, the problem is not important. Just document this corner case.

This is also why it could be possible with a result like "--cached --cached". I see no point in wasting readability (and cpu cycles!) to handle this situation.  

if (from == StagedOpt && to == StagedOpt)
{
  to = String.Empty;
}